---
title: "Differentially Methylated Gene Analysis"
author: "Yaamini Venkataraman"
date: "8/19/2019"
output: github_document
---

In this script, I will identify genes that are differentially methylated in response to *C. virginica* exposure to experimental ocean acidification conditions. I am roughly replicating an analysis from [this paper](https://advances.sciencemag.org/content/4/6/eaar8028) and [this preprint](https://www.biorxiv.org/content/early/2018/02/21/269076). .I will annotate coverage output from `bismark` with gene information, calculate median methylation for each gene, then perform t-tests for each gene. For differentially methylated genes (DMG), I will conduct a functional enrichment with GO-MWU.

**NOTE: MOVE THIS R MARKDOWN FILE INTO THE `2019-08-14-Differentially-Methylated-Genes` FOLDER BEFORE RUNNING.**

# Set up R Markdown document

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Obtain session information

```{r}
sessionInfo()
```

# Create annotated gene list

## Add gene track

```{r}
geneList <- read.delim("../genome-feature-tracks/C_virginica-3.0_Gnomon_gene_sorted_yrv.gff3", header = FALSE) #Import gene gff3
colnames(geneList) <- c("gene-chr", "Gnomon", "gene", "gene-start", "gene-end", "V6", "V7", "V8", "V9") #Rename columns
head(geneList) #Confirm import
```

```{r}
geneList <- geneList[,-c(6:8)] #Remove columns 6-8
head(geneList) #Confirm changes
```

### Import gene ID

```{r}
geneDescriptions <- read.delim2("../2018-12-02-Gene-Enrichment-Analysis/2019-06-20-Gene-Descriptions.txt", header = FALSE, col.names = c("gene-number", "gene-ID", "V3", "V4", "V5", "V6", "V7", "V8", "V9")) #Import document with gene descrptions. Name 9 columns so all rows are imported with 9 columns
head(geneDescriptions) #Confirm import
```

```{r}
geneID <- geneDescriptions$gene.ID #Isolate gene ID column
geneID <- gsub("Dbxref=GeneID:", "", geneID) #Remove extraneous characters
head(geneID) #Confirm changes
```

### Merge gene location and gene ID

```{r}
geneList <- data.frame(geneList, geneID) #Replace geneList dataframe with geneList and geneID
geneList <- geneList[,-6] #Remove extra column
colnames(geneList) <- c("gene-chr", "Gnomon", "gene", "gene-start", "gene-end", "gene-ID") #Rename columns to match previous naming conventions
head(geneList) #Confirm merge
```

## Add mRNA track

```{r}
mRNAList <- read.delim("../genome-feature-tracks/C_virginica-3.0_Gnomon_mRNA_yrv.gff3", header = FALSE) #Import mRNA track
mRNAList <- mRNAList[,-c(6,8)] #Remove extra columns
colnames(mRNAList) <- c("mRNA-chr", "Gnomon", "mRNA", "mRNA-start", "mRNA-end", "mRNA-strand", "V9") #Rename columns
head(mRNAList) #Confirm import
```

### Import mRNA descriptions

```{r}
mRNADescriptions <- read.delim2("../2018-12-02-Gene-Enrichment-Analysis/2019-06-20-mRNA-Descriptions.txt", header = FALSE, col.names = paste("V", 1:78, sep = "")) #Import document with gene descrptions. Name 78 columns so all rows are imported with 78 columns
head(mRNADescriptions) #Confirm import
```

```{r}
geneIDmRNAList <- mRNADescriptions$V3 #Isolate gene ID column
geneIDmRNAList <- gsub("Dbxref=GeneID:", "", geneIDmRNAList) #Remove extraneous characters
head(geneIDmRNAList) #Confirm changes
```

```{r}
genbankIDmRNAList <- mRNADescriptions$V4 #Isolate Genbank ID
genbankIDmRNAList <- gsub("Genbank:", "", genbankIDmRNAList) #Remove extraneous characters
head(genbankIDmRNAList) #Confirm changes
```

### Import isolated mRNA products

```{r}
mRNAProducts <- read.delim2("../2018-12-02-Gene-Enrichment-Analysis/2019-06-20-mRNA-Products.txt", header = FALSE, col.names = paste("V", 1:16, sep = "")) #Import document with gene descrptions. Name 16 columns so all rows are imported with 16 columns.
head(mRNAProducts) #Confirm import
```

```{r}
mRNAProducts <- mRNAProducts$V1 #Extract the first column with product information
```

### Reformat mRNA List

```{r}
mRNAList <- data.frame(mRNAList, geneIDmRNAList, genbankIDmRNAList, mRNAProducts) #Replace mRNAList dataframe with mRNAList, geneIDmRNAList, genbankIDmRNAList, and mRNAProducts
mRNAList <- mRNAList[,-c(6:7)] #Remove extra columns
colnames(mRNAList) <- c("mRNA-chr", "Gnomon", "mRNA", "mRNA-start", "mRNA-end", "gene-ID", "Genbank", "Product") #Rename columns to match previous naming conventions
head(mRNAList) #Confirm merge
```

## Merge gene and mRNA tracks

```{r}
geneAnnotations <- unique(merge(x = geneList, y = mRNAList, by = "gene-ID")) #Merge gene and mRNA lists by gene-ID column. I will only have genes with mRNA in the final dataframe. Additionally, keep only unique entries (remove any merging artifacts)
length(geneAnnotations$`gene-ID`) #There are 60,201 mRNA, so this is a good sign!
```

```{r}
geneAnnotations <- geneAnnotations[,c(2, 5:6, 10:11, 1, 12, 13)] #Reorganize columns
head(geneAnnotations) #Confirm reorganization
```

## Retain the longest mRNA for each gene

When looking at differentially methylated genes, Liew et al. used the longest mRNA for each gene. I will do the same thing.

```{r}
geneAnnotations$mRNALength <- as.numeric(geneAnnotations$`mRNA-end` - geneAnnotations$`mRNA-start`) #Calculate the length of each mRNA and save in a new column
head(geneAnnotations) #Confirm length calculation
```

```{r}
attach(geneAnnotations) #Attach dataframe
geneAnnotations <- geneAnnotations[order(`gene-chr`, `gene-start`, -mRNALength),] #Sort by gene chromosome, then gene start position, and finally mRNA length. The minus sign indicates the dataframe should be sorted in descending order (longer mRNA length first)
detach(geneAnnotations) #Detach dataframe
head(geneAnnotations) #Confirm sorting
```

```{r}
write.table(geneAnnotations, "2019-08-14-All-Annotated-Gene-and-mRNA.txt", quote = FALSE, row.names = FALSE) #Save ordered file
```

```{bash}
#For the 6th field [$6] (gene ID), remove all duplicate entries in a csv file (-F","). These entries are saved and can be extracted as a different output if needed. 
#Save output to a new file
awk -F"" '!x[$6]++' 2019-08-14-All-Annotated-Gene-and-mRNA.txt \
> ../analyses/2019-08-14-Differentially-Methylated-Genes/2019-08-14-All-Annotated-Gene-and-mRNA-Unique.csv
```

```{bash}
head ../analyses/2019-08-14-Differentially-Methylated-Genes/2019-08-14-All-Annotated-Gene-and-mRNA-Unique.csv #Check file
```

```{r}
geneAnnotationsCondensed <- read.csv("2019-08-14-All-Annotated-Gene-and-mRNA-Unique.csv", header = TRUE) #Import condensed information
length(geneAnnotationsCondensed$gene.chr) #Number of mRNA from unique genes.
```

## Add Uniprot and GOterm information

```{r}
blastResults <- read.delim("../2018-12-02-Gene-Enrichment-Analysis/2018-09-11-Transcript-Uniprot-blastx-codeIsolated.txt", header = FALSE, sep = "\t", dec = ".") #Import blast results
blastResults <- blastResults[,c(1, 3)] #Keep only relevant columns
colnames(blastResults) <- c("Genbank", "Uniprot") #Rename columns
head(blastResults) #Confirm column headers
```

```{r}
geneAnnotationsCondensedUniprot <- unique(merge(x = geneAnnotationsCondensed, y = blastResults, by = "Genbank", all.x = TRUE, sort = FALSE)) #Add Uniprot Accession code to gene annotations. Keep all rows even if they don't have matching Uniprot codes. Do not resort the output. Keep only unique entries (remove merging artifacts)
head(geneAnnotationsCondensedUniprot) #Confirm merge
```

```{r}
uniprotGOTerms <- read.delim("../2018-12-02-Gene-Enrichment-Analysis/uniprot-reviewed_yes.tab", header = TRUE) #Import tab-delimited file with header
uniprotGOTerms <- uniprotGOTerms[,c(1,8)] #Retain Entry and GOterm columns
colnames(uniprotGOTerms) <- c("Uniprot", "GO") #Rename columns
head(uniprotGOTerms) #Confirm import
```

```{r}
geneAnnotationsCondensedUniprotGOterms <- unique(merge(x = geneAnnotationsCondensedUniprot, y = uniprotGOTerms, by = "Uniprot", all.x = TRUE)) #Add GOterms to gene annotations. Keep all rows even if they don't have matching Uniprot codes. Keep only unique entries (remove merging artifacts)
head(geneAnnotationsCondensedUniprotGOterms, n = 2) #Confirm merge. The output is sorted by Uniprot codes, so I'll need to resort. It would have been sorted by Uniprot code order even if I had used sort = FALSE.
```

```{r}
geneAnnotationsCondensedUniprotGOterms <- geneAnnotationsCondensedUniprotGOterms[,c(3:8, 2, 10, 9, 1, 11)] #Reorganize columns
attach(geneAnnotationsCondensedUniprotGOterms) #Attach dataframe
geneAnnotationsCondensedUniprotGOterms <- geneAnnotationsCondensedUniprotGOterms[order(gene.chr, gene.start),] #Sort by gene chromosome, then start position
detach(geneAnnotationsCondensedUniprotGOterms) #Detatch dataframe
head(geneAnnotationsCondensedUniprotGOterms, n = 5) #Confirm reorganization and sorting
```

I will now save the gene information as a new file. This information will be used to annotate coverage files.

```{r}
write.csv(geneAnnotationsCondensedUniprotGOterms, "2019-08-14-Annotated-Gene-Longest-mRNA-Uniprot-GOterms.csv", quote = FALSE, row.names = FALSE) #Save file
```

```{r}
write.table(geneAnnotationsCondensedUniprotGOterms, "2019-08-14-Annotated-Gene-Longest-mRNA-Uniprot-GOterms.txt", sep = "\t", col.names = FALSE, row.names = FALSE, quote = FALSE) #Also saving a tab-delimited version
```

# Annotate coverage files

## Download 5x coverage files

Liew et al. used raw coverage files from `bismark` for their analysis, but I will use manipulated coverage files. In [this Jupyter notebook](https://github.com/fish546-2018/yaamini-virginica/blob/master/notebooks/2019-03-07-Generating-Coverage-Tracks.ipynb), I filtered loci with 5x coverage and saved that information in bedGraphs. These bedGraphs will be used in the DMG analysis.

```{bash, echo = FALSE}
wget -r -l1 --no-parent -A_5x.bedgraph https://gannet.fish.washington.edu/spartina/2018-10-10-project-virginica-oa-Large-Files/2019-03-07-Yaamini-Virginica-Repository/analyses/2019-03-07-IGV-Verification/ #Download 5x bedgraphs from gannet
```

```{bash}
mv gannet.fish.washington.edu/spartina/2018-10-10-project-virginica-oa-Large-Files/2019-03-07-Yaamini-Virginica-Repository/analyses/2019-03-07-IGV-Verification/* . #Move files to analysis directory. Where they are stored mimics file structure on gannet
```

```{bash}
rm -r gannet.fish.washington.edu #Remove gannet directory
```

```{bash}
head zr2096_1_s1_R1_val_1_bismark_bt2_pe.deduplicated.bismark.cov_5x.bedgraph #Columns: chr, start, end, percent methylation. Tab-delimited.
```

## Create new files without percent methylation information

To match loci with the gene they're in, I'm going to use `intersectBed` from `bedtools`. Before I can use files with `intersectBed`, I need to remove the last column from the coverage files, percent methylation. This is because `intersectBed` will only accept chromosome and positions, and will not accept columns with numbers that are not positions.

```{bash}
#For each file that ends in bedgraph
#Print the first three columns, tab-delimited, and save the columns in a new file that has the same base name and ends in posOnly
for f in *bedgraph
do
  awk '{print $1"\t"$2"\t"$3}' ${f} > ${f}.posOnly
done
```

```{bash}
find *posOnly #Confirm files were created
```

```{bash}
head zr2096_5_s1_R1_val_1_bismark_bt2_pe.deduplicated.bismark.cov_5x.bedgraph.posOnly #Confirm file formatting
```

## Run `intersectBed`

```{bash}
#For each file that ends in posOnly
#Use intersectBed to find where loci and genes intersect, allowing loci to be mapped to annotated genes
#wb: Print all lines in the second file
#a: file that ends in posOnly
#b: annotated gene list
#Save output in a new file that has the same base name and ends in -Annotated.txt
for f in *posOnly
do
  /Users/Shared/bioinformatics/bedtools2/bin/intersectBed \
  -wb \
  -a ${f} \
  -b 2019-08-14-Annotated-Gene-Longest-mRNA-Uniprot-GOterms.txt \
  > ${f}-Annotated.txt
done
```

```{bash}
find *Annotated.txt
```

```{bash}
head zr2096_1_s1_R1_val_1_bismark_bt2_pe.deduplicated.bismark.cov_5x.bedgraph.posOnly-Annotated.txt #Confirm file is annotated
```

## Add back percent methylation information

After annotating coverage files with `intersectBed`, I need to add percent methylation information back to the annotated files. I need the percent methylation data to calculate median methylation by gene. I could import 20 files into R, or I could use a series of `bash` commands.

The original files aren't tab-delimited while the others are, so I need to get the coverage files in a tab-delimited format so my output will be tab-delimited.

```{bash}
#For each file that ends in bedgraph (the original coverage files)
#Print the first three columns (chr, start, end) with dashes ("-") in between. Then, print the rest of the columns (chr, start, end, percentMeth) in the file with tabs ("\t") inbetween.
#Save the file with the base name + .sorted
for f in *bedgraph
do
  awk '{print $1"-"$2"-"$3"\t"$1"\t"$2"\t"$3"\t"$4}' ${f} \
  | sort -k1,1 \
  > ${f}.sorted
done
```

```{bash}
find *bedgraph.sorted
```

```{bash}
head zr2096_1_s1_R1_val_1_bismark_bt2_pe.deduplicated.bismark.cov_5x.bedgraph.sorted
```

```{bash}
#For each file that ends in Annotated.txt
#Print the first three columns (chr, start, end) with dashes in between, then the rest of the columns in the file
#Save the file with the base name + .sorted
for f in *Annotated.txt
do
  awk '{print $1"-"$2"-"$3"\t"$0}' ${f} \
  | sort -k1,1 \
  > ${f}.sorted
done
```

```{bash}
find *Annotated.txt.sorted
```

```{bash}
head zr2096_1_s1_R1_val_1_bismark_bt2_pe.deduplicated.bismark.cov_5x.bedgraph.posOnly-Annotated.txt.sorted
```

```{bash}
#For each file that ends in bedgraph
#Join 2 files using the first column. The files and tab-delimited and the output should also be tab-delimited
#The first file ends with .bedgraph.sorted
#The second file ends with .bedgraph.posOnly.Annotated.txt.sorted
#Add .annotated.percentMeth to the base name of the output file
for f in *bedgraph
do
  join -j 1 -t $'\t' \
  ${f}.sorted \
  ${f}.posOnly-Annotated.txt.sorted \
  > ${f}.annotated-percentMeth
done
```

```{bash}
find *annotated-percentMeth
```

```{bash}
head zr2096_1_s1_R1_val_1_bismark_bt2_pe.deduplicated.bismark.cov_5x.bedgraph.annotated-percentMeth
```

```{bash}
for f in *Annotated.txt
do
  wc -l ${f}
done
```

```{bash}
for f in *percentMeth
do
  wc -l ${f}
done
```

## Add column names to files

```{bash}
#For each file that ends in percentMeth
#List tab-delimited column headers
#Add the file after the headers
#Save output as a new file with the base name + ".header"
for f in *percentMeth
do
  echo -e "key\tchr\tstart\tend\tpercentMeth\tchr2\tstart2\tend2\tgene-chr\tgene-start\tgene-end\tmRNA-start\tmRNA-end\tgeneID\tGenbank\tmRNALength\tProduct\tUniprot\tGO" \
  | cat - ${f} \
  > ${f}.header
done
```

```{bash}
find *header
```

```{bash}
head zr2096_1_s1_R1_val_1_bismark_bt2_pe.deduplicated.bismark.cov_5x.bedgraph.annotated-percentMeth.header
```

# Calculate median percent methylation for each sample

I can do this within R, so the first step is to import the files.

```{r}
filesToImport <- list.files(pattern = "*header") #Create a file list for all 10 files to import
list2env(lapply(setNames(filesToImport,
                         make.names(gsub("_s1_R1_val_1_bismark_bt2_pe.deduplicated.bismark.cov_5x.bedgraph.annotated-percentMeth.header", "", filesToImport))),
                read.delim),
         envir = .GlobalEnv) #Import files to the .GlobalEnv with list2env. Use lapply to setNames of the files by taking all the common parts of their names out. Read files with read.delim (can only use defaults = header = TRUE, sep = "\t"). Files will be named zr2096_#
head(zr2096_1) #Confirm import
```

```{r}
#Remove extraneous columns...there has to be a better way to do this in R
zr2096_1 <- zr2096_1[,-c(1, 6:8)]
zr2096_2 <- zr2096_2[,-c(1, 6:8)]
zr2096_3 <- zr2096_3[,-c(1, 6:8)]
zr2096_4 <- zr2096_4[,-c(1, 6:8)]
zr2096_5 <- zr2096_5[,-c(1, 6:8)]
zr2096_6 <- zr2096_6[,-c(1, 6:8)]
zr2096_7 <- zr2096_7[,-c(1, 6:8)]
zr2096_8 <- zr2096_8[,-c(1, 6:8)]
zr2096_9 <- zr2096_9[,-c(1, 6:8)]
zr2096_10 <- zr2096_10[,-c(1, 6:8)]
head(zr2096_1) #Confirm column removal
```

```{r}
#Calculate median percent methylation by gene for each sample
zr2096_1_PercentMeth <- aggregate(percentMeth ~ geneID, data = zr2096_1, FUN = median)
zr2096_2_PercentMeth <- aggregate(percentMeth ~ geneID, data = zr2096_2, FUN = median)
zr2096_3_PercentMeth <- aggregate(percentMeth ~ geneID, data = zr2096_3, FUN = median)
zr2096_4_PercentMeth <- aggregate(percentMeth ~ geneID, data = zr2096_4, FUN = median)
zr2096_5_PercentMeth <- aggregate(percentMeth ~ geneID, data = zr2096_5, FUN = median)
zr2096_6_PercentMeth <- aggregate(percentMeth ~ geneID, data = zr2096_6, FUN = median)
zr2096_7_PercentMeth <- aggregate(percentMeth ~ geneID, data = zr2096_7, FUN = median)
zr2096_8_PercentMeth <- aggregate(percentMeth ~ geneID, data = zr2096_8, FUN = median)
zr2096_9_PercentMeth <- aggregate(percentMeth ~ geneID, data = zr2096_9, FUN = median)
zr2096_10_PercentMeth <- aggregate(percentMeth ~ geneID, data = zr2096_10, FUN = median)
head(zr2096_1_PercentMeth) #Confirm percent methylation calculation
```

```{r}
#Merge sample percent methylation information so that all geneIDs have data
fullPercentMeth <- merge(x = zr2096_1_PercentMeth, y = zr2096_2_PercentMeth, by = "geneID")
fullPercentMeth <- merge(x = fullPercentMeth, y = zr2096_3_PercentMeth, by = "geneID")
fullPercentMeth <- merge(x = fullPercentMeth, y = zr2096_4_PercentMeth, by = "geneID")
fullPercentMeth <- merge(x = fullPercentMeth, y = zr2096_5_PercentMeth, by = "geneID")
fullPercentMeth <- merge(x = fullPercentMeth, y = zr2096_6_PercentMeth, by = "geneID")
fullPercentMeth <- merge(x = fullPercentMeth, y = zr2096_7_PercentMeth, by = "geneID")
fullPercentMeth <- merge(x = fullPercentMeth, y = zr2096_8_PercentMeth, by = "geneID")
fullPercentMeth <- merge(x = fullPercentMeth, y = zr2096_9_PercentMeth, by = "geneID")
fullPercentMeth <- merge(x = fullPercentMeth, y = zr2096_10_PercentMeth, by = "geneID")
length(fullPercentMeth$geneID) #Number of gene IDs in combined dataset.
head(fullPercentMeth) #Confirm merge. Column names are replicated but the columns are from samples 1-10.
```

```{r}
#Create individual dataframes
zr2096_1_PercentMethFull <- fullPercentMeth[,c(1,2)]
zr2096_2_PercentMethFull <- fullPercentMeth[,c(1,3)]
zr2096_3_PercentMethFull <- fullPercentMeth[,c(1,4)]
zr2096_4_PercentMethFull <- fullPercentMeth[,c(1,5)]
zr2096_5_PercentMethFull <- fullPercentMeth[,c(1,6)]
zr2096_6_PercentMethFull <- fullPercentMeth[,c(1,7)]
zr2096_7_PercentMethFull <- fullPercentMeth[,c(1,8)]
zr2096_8_PercentMethFull <- fullPercentMeth[,c(1,9)]
zr2096_9_PercentMethFull <- fullPercentMeth[,c(1,10)]
zr2096_10_PercentMethFull <- fullPercentMeth[,c(1,11)]
head(zr2096_1_PercentMethFull) #Confirm dataframe creation
```

```{r}
#Change column names
colnames(zr2096_1_PercentMethFull)[2] <- "percentMeth"
colnames(zr2096_2_PercentMethFull)[2] <- "percentMeth"
colnames(zr2096_3_PercentMethFull)[2] <- "percentMeth"
colnames(zr2096_4_PercentMethFull)[2] <- "percentMeth"
colnames(zr2096_5_PercentMethFull)[2] <- "percentMeth"
colnames(zr2096_6_PercentMethFull)[2] <- "percentMeth"
colnames(zr2096_7_PercentMethFull)[2] <- "percentMeth"
colnames(zr2096_8_PercentMethFull)[2] <- "percentMeth"
colnames(zr2096_9_PercentMethFull)[2] <- "percentMeth"
colnames(zr2096_10_PercentMethFull)[2] <- "percentMeth"
head(zr2096_1_PercentMethFull) #Confirm name change
```

```{r}
#Add sample number to each dataframe
zr2096_1_PercentMethFull$sampleNumber <- rep("sample_1", times = length(zr2096_1_PercentMethFull$geneID))
zr2096_2_PercentMethFull$sampleNumber <- rep("sample_2", times = length(zr2096_2_PercentMethFull$geneID))
zr2096_3_PercentMethFull$sampleNumber <- rep("sample_3", times = length(zr2096_3_PercentMethFull$geneID))
zr2096_4_PercentMethFull$sampleNumber <- rep("sample_4", times = length(zr2096_4_PercentMethFull$geneID))
zr2096_5_PercentMethFull$sampleNumber <- rep("sample_5", times = length(zr2096_5_PercentMethFull$geneID))
zr2096_6_PercentMethFull$sampleNumber <- rep("sample_6", times = length(zr2096_6_PercentMethFull$geneID))
zr2096_7_PercentMethFull$sampleNumber <- rep("sample_7", times = length(zr2096_7_PercentMethFull$geneID))
zr2096_8_PercentMethFull$sampleNumber <- rep("sample_8", times = length(zr2096_8_PercentMethFull$geneID))
zr2096_9_PercentMethFull$sampleNumber <- rep("sample_9", times = length(zr2096_9_PercentMethFull$geneID))
zr2096_10_PercentMethFull$sampleNumber <- rep("sample_10", times = length(zr2096_10_PercentMethFull$geneID))
head(zr2096_5_PercentMethFull) #Confirm column addition
```

```{r}
#Add treatment indication to each dataframe
zr2096_1_PercentMethFull$treatment <- rep("Ambient", times = length(zr2096_1_PercentMethFull$geneID))
zr2096_2_PercentMethFull$treatment <- rep("Ambient", times = length(zr2096_2_PercentMethFull$geneID))
zr2096_3_PercentMethFull$treatment <- rep("Ambient", times = length(zr2096_3_PercentMethFull$geneID))
zr2096_4_PercentMethFull$treatment <- rep("Ambient", times = length(zr2096_4_PercentMethFull$geneID))
zr2096_5_PercentMethFull$treatment <- rep("Ambient", times = length(zr2096_5_PercentMethFull$geneID))
zr2096_6_PercentMethFull$treatment <- rep("Treatment", times = length(zr2096_6_PercentMethFull$geneID))
zr2096_7_PercentMethFull$treatment <- rep("Treatment", times = length(zr2096_7_PercentMethFull$geneID))
zr2096_8_PercentMethFull$treatment <- rep("Treatment", times = length(zr2096_8_PercentMethFull$geneID))
zr2096_9_PercentMethFull$treatment <- rep("Treatment", times = length(zr2096_9_PercentMethFull$geneID))
zr2096_10_PercentMethFull$treatment <- rep("Treatment", times = length(zr2096_10_PercentMethFull$geneID))
head(zr2096_1_PercentMethFull) #Confirm column addition
```

```{r}
fullPercentMethExpanded <- rbind(zr2096_1_PercentMethFull,
                                 zr2096_2_PercentMethFull,
                                 zr2096_3_PercentMethFull,
                                 zr2096_4_PercentMethFull,
                                 zr2096_5_PercentMethFull,
                                 zr2096_6_PercentMethFull,
                                 zr2096_7_PercentMethFull,
                                 zr2096_8_PercentMethFull,
                                 zr2096_9_PercentMethFull,
                                 zr2096_10_PercentMethFull) #Create new dataframe with all sample information
head(fullPercentMethExpanded) #Confirm dataframe creation
```

```{r}
write.csv(fullPercentMethExpanded, "2019-08-22-Gene-Median-Percent-Methylation-By-Sample.csv", row.names = FALSE, quote = FALSE) #Save file
```

# Calculate methylation differences between treatments

If I have any differentially methylated genes, I want methylation differences for each gene by treatment to use with gene enrichment. To do this, I'm going to calculate median percent methylation by treatment, then subtract ambient values from treatment values.

```{r}
ambientSamples <- subset(fullPercentMethExpanded, subset = fullPercentMethExpanded$treatment == "Ambient") #Subset ambient samples
treatmentSamples <- subset(fullPercentMethExpanded, subset = fullPercentMethExpanded$treatment == "Treatment") #Subset treatment samples
head(treatmentSamples) #Confirm subset
```

```{r}
ambientSamplesPercentMeth <- aggregate(percentMeth ~ geneID, data = ambientSamples, FUN = median) #Calculate median methylation by gene for ambient samples
treatmentSamplesPercentMeth <- aggregate(percentMeth ~ geneID, data = treatmentSamples, FUN = median) #Calculate median methylation by gene for treatment samples
head(treatmentSamplesPercentMeth) #Confirm calculations
```

```{r}
fullMethDiff <- data.frame("geneID" = ambientSamplesPercentMeth$geneID,
                           "meth.diff" = (treatmentSamplesPercentMeth$percentMeth - ambientSamplesPercentMeth$percentMeth)) #Create a new dataframe with geneID and methylation differences (treatment - ambient)
head(fullMethDiff) #Confirm dataframe creation
```

# Identifying differentially methylated genes

## Check normality

I want to do a t-test, so I should first check normality first.
`
```{r}
hist(fullPercentMethExpanded$percentMeth) #Check normality
```

This clearly isn't a normal distributions. I need to do some corrections to deal with the left-skewness.

```{r}
fullPercentMethExpanded$logPercentMeth <- log(fullPercentMethExpanded$percentMeth) #Perform log-transformation
fullPercentMethExpanded$sqrtPercentMeth <- sqrt(fullPercentMethExpanded$percentMeth) #Perform square root transformation
fullPercentMethExpanded$crtPercentMeth <- (fullPercentMethExpanded$percentMeth)^(1/3) #Perform cube root transformation
```

```{r}
#Check normality
hist(fullPercentMethExpanded$logPercentMeth) #Log transformation
hist(fullPercentMethExpanded$sqrtPercentMeth) #Square root transformation
hist(fullPercentMethExpanded$crtPercentMeth) #Cube root transformation
```

Well, none of these transformations are very helpful! The MBD bias is pretty evident, even in the transformed data. I'll condut a non-parametric test instead.

## Non-parametric test

### Format for statistical testing

```{r}
attach(fullPercentMethExpanded) #Attach dataframe
fullPercentMethExpanded <- fullPercentMethExpanded[order(geneID, treatment),] #Sort by gene ID, then by treatment
detach(fullPercentMethExpanded) #Detach dataframe
head(fullPercentMethExpanded) #Confirm sorting
```


```{r}
ambientPercentMethFull <- subset(fullPercentMethExpanded, subset = fullPercentMethExpanded$treatment == "Ambient") #Subset ambient samples
treatmentPercentMethFull <- subset(fullPercentMethExpanded, subset = fullPercentMethExpanded$treatment == "Treatment") #Subset treatment samples
head(ambientPercentMethFull) #Confirm subset
head(treatmentPercentMethFull) #Confirm subset
```

### Conduct Kruskal-Wallis test

```{r}
KWTestResults <- data.frame("geneID" = unique(ambientPercentMethFull$geneID),
                            "chi-squared" = rep(0, times = length(unique(ambientPercentMethFull$geneID))),
                            "df" = rep(0, times = length(unique(ambientPercentMethFull$geneID))),
                            "p-value" = rep(0, times = length(unique(ambientPercentMethFull$geneID)))) #Create a new dataframe to store results of statistical test
head(KWTestResults) #Confirm dataframe creation
```

```{r}
#For each unique geneID
#Conduct a KW test for a set of 5 rows from each dataframe (5 ambient and 5 treatment median % methylation values for the same gene)
#Store the chi-squared statistic in the ith row
#Store the df in the ith row
#Store the p.value in the ith row
for (i in 1:length(KWTestResults$geneID)) {
  temp <- kruskal.test(list(ambientPercentMethFull$percentMeth[((5*i)-4):(5*i)], treatmentPercentMethFull$percentMeth[((5*i)-4):(5*i)]))
  KWTestResults$chi.squared[i] <- temp$statistic
  KWTestResults$df[i] <- temp$parameter
  KWTestResults$p.value[i] <- temp$p.value
}
head(KWTestResults) #Confirm statistical tests worked
```

```{r}
KWTestResults$adj.p <- p.adjust(KWTestResults$p.value, method = "BH") #Control the false discovery rate using a Benjamini-Hochberg correction
range(KWTestResults$adj.p, na.rm = TRUE) #Not counting NAs, there are no genes with significant differential methylation. No need to annotate genes! 
```

```{r}
write.csv(KWTestResults, "2019-08-14-DMG-KW-Test-Results.csv", quote = FALSE, row.names = FALSE) #Save file
```

### Annotate results

```{r}
colnames(geneAnnotationsCondensedUniprotGOterms)[6] <- "geneID" #Rename column so it's consistent with KW results
colnames(geneAnnotationsCondensedUniprotGOterms) #Confirm column was renamed
```

```{r}
uncorrGenes <- unique(merge(x = KWTestResults, y = geneAnnotationsCondensedUniprotGOterms, by = "geneID")) #Annotate genes and remove any merging artifacts
length(uncorrGenes$geneID) == length(KWTestResults$geneID) #Check that gene IDs are not duplicated
head(uncorrGenes) #Confirm merge
```

```{r}
uncorrGenes <- unique(merge(x = uncorrGenes, y = fullMethDiff, by = "geneID")) #Add meth.diff information
length(uncorrGenes$geneID) == length(KWTestResults$geneID) #Check that gene IDs are not duplicated
head(uncorrGenes) #Confirm merge
```

```{r}
write.csv(uncorrGenes, "2019-08-14-KW-Test-Results-Annotated.csv", quote = FALSE, row.names = FALSE) #Save file
```
